### ISSUE TITLE
Recursive data types

### ISSUE THREAD

**Issue by Profpatsch** [dhall-lang/dhall-lang#3]
I suspect being not turing complete, the following is not possible;

Is it possible to represent a recursive type in dhall?
e.g. a `Node` union that references itself:

Node.dhall:
```haskell
< Plain : Text
| Heading : { level : Natural
            , heading : Text }
| Listing : List Node
>
```

akin to

```haskell
data Node
  = Plain Text
  | Heading { level :: Int, heading :: Text }
  | Listing : [Node]
```

Recursive imports are not allowed, but maybe I’m missing something.

**Comment by Profpatsch** [comment]
My use case is specifying a small eDSL for (reactjs) slides in dhall and translating them to HTML afterwards.

**Comment by Gabriel439** [comment]
It is definitely possible!  Although the language doesn't provide built-in support for recursive types, you can still encode them within the language using a mechanical translation process, like this:

```
$ cat Node 
    let Node : Type =
            ∀(Node : Type)
        →   ∀(  Recurse
            :   < Plain   : Text
                | Heading : { level : Natural, heading : Text }
                | Listing : List Node
                >
            →   Node
            )
        →   Node
in  Node
```

```
$ cat Plain
    let Plain : Text → ./Node =
            λ(x : Text)
        →   λ(Node : Type)
        →   λ(  Recurse
            :   < Plain   : Text
                | Heading : { level : Natural, heading : Text }
                | Listing : List Node
                >
            →   Node
            )
        →   Recurse
            < Plain   = x
            | Heading : { level : Natural, heading : Text }
            | Listing : List Node
            >
in  Plain
```

```
$ cat Heading 
    let Heading : { level : Natural, heading : Text } → ./Node =
            λ(x : { level : Natural, heading : Text })
        →   λ(Node : Type)
        →   λ(  Recurse
            :   < Plain   : Text
                | Heading : { level : Natural, heading : Text }
                | Listing : List Node
                >
            →   Node
            )
        →   Recurse
            < Heading = x
            | Plain   : Text
            | Listing : List Node
            >
in  Heading
```

```
$ cat Listing 
    let map = https://ipfs.io/ipfs/QmQ8w5PLcsNz56dMvRtq54vbuPe9cNnCCUXAQp6xLc6Ccx/Prelude/List/map

in  let Listing : List ./Node → ./Node =
            λ(x : List ./Node)
        →   λ(Node : Type)
        →   λ(  Recurse
            :   < Plain   : Text
                | Heading : { level : Natural, heading : Text }
                | Listing : List Node
                >
            →   Node
            )
        →   let f : ./Node → Node =
                λ(x : ./Node) → x Node Recurse
        in  Recurse
            < Listing = map ./Node Node f x
            | Plain   : Text
            | Heading : { level : Natural, heading : Text }
            >
in  Listing
```

... and here is an example of how a user would use this:

```
$ cat example
./Listing
[   ./Plain "ABC"
,   ./Listing
    [   ./Heading { level = +0, heading = "foo" }
    ,   ./Listing
        [   ./Heading { level = +1, heading = "bar" }
        ]
    ,   ./Plain "DEF"
    ]
,   ./Plain "GHI"
]
```

... which normalizes to:

```
$ dhall <<< './example'
∀(Node : Type) → ∀(Recurse : < Heading : { heading : Text, level : Natural } | Listing : List Node | Plain : Text > → Node) → Node

λ(Node : Type) → λ(Recurse : < Heading : { heading : Text, level : Natural } | Listing : List Node | Plain : Text > → Node) → Recurse < Listing = [Recurse < Plain = "ABC" | Heading : { heading : Text, level : Natural } | Listing : List Node >, Recurse < Listing = [Recurse < Heading = { heading = "foo", level = +0 } | Listing : List Node | Plain : Text >, Recurse < Listing = [Recurse < Heading = { heading = "bar", level = +1 } | Listing : List Node | Plain : Text >] : List Node | Heading : { heading : Text, level : Natural } | Plain : Text >, Recurse < Plain = "DEF" | Heading : { heading : Text, level : Natural } | Listing : List Node >] : List Node | Heading : { heading : Text, level : Natural } | Plain : Text >, Recurse < Plain = "GHI" | Heading : { heading : Text, level : Natural } | Listing : List Node >] : List Node | Heading : { heading : Text, level : Natural } | Plain : Text >
```

... or formatted for readability:

```
    λ(Node : Type)
→   λ(Recurse : < Heading : { heading : Text, level : Natural } | Listing : List Node | Plain : Text > → Node)
→   Recurse < Listing =
        [ Recurse < Plain = "ABC" | Heading : { heading : Text, level : Natural } | Listing : List Node >
        , Recurse < Listing =
            [ Recurse < Heading = { heading = "foo", level = +0 } | Listing : List Node | Plain : Text >
            , Recurse < Listing =
                [ Recurse < Heading = { heading = "bar", level = +1 } | Listing : List Node | Plain : Text >
                ] : List Node | Heading : { heading : Text, level : Natural } | Plain : Text >
            , Recurse < Plain = "DEF" | Heading : { heading : Text, level : Natural } | Listing : List Node >
            ] : List Node | Heading : { heading : Text, level : Natural } | Plain : Text >
        , Recurse < Plain = "GHI" | Heading : { heading : Text, level : Natural } | Listing : List Node >
        ] : List Node | Heading : { heading : Text, level : Natural } | Plain : Text >
```

**Comment by Gabriel439** [comment]
Actually, thinking about this more, it might be easier to just hand-write the latter form.  In other words, just introduce the two bound variables:

```haskell
    λ(Node : Type)
→   λ(Recurse : < Heading : { heading : Text, level : Natural } | Listing : List Node | Plain : Text > → Node)
→   ...
```

... and then wrap each level of nesting in `Recurse` when defining the nested data type

**Comment by Gabriel439** [comment]
I forgot to mention that if you don't care about using anonymous unions specifically, there is a less verbose encoding:

```haskell
    λ(Node : Type)
→   λ(Plain : Text → Node)
→   λ(Heading : { level : Natural, heading : Text } → Node)
→   λ(Listing : List Node → Node)
→   Listing
    [   Plain "ABC"
    ,   Listing
        [   Heading { level = +0, heading = "foo" }
        ,   Listing
            [   Heading { level = +1, heading = "bar" }
            ]
        ,   Plain "DEF"
        ]
    ,   Plain "GHI"
    ]
```

That's probably easier to read and directly manipulate

**Comment by Profpatsch** [comment]
Oh! That looks like a GADT.

**Comment by Profpatsch** [comment]
Wait; how can I create something `→ Node` then? e.g. what would a valid definition of `Plain` be?

**Comment by Gabriel439** [comment]
So the type of `./Node` would be:

```haskell
-- ./Node

    ∀(Node : Type)
→   ∀(Plain : Text → Node)
→   ∀(Heading : { level : Natural, heading : Text } → Node)
→   ∀(Listing : List Node → Node)
→   Node
```

... and then you would implement `Plain` as:

```haskell
-- ./Plain

let Plain : Text → ./Node =
        λ(x : Text)
    →   λ(Node : Type)
    →   λ(Plain : Text → Node)
    →   λ(Heading : { level : Natural, heading : Text } → Node)
    →   λ(Listing : List Node → Node)
    →   Plain x

in  Plain
```

Marshaling it into Haskell data via `Interpret` would be trickier, but still possible.  You wouldn't be able to do it by just deriving `Generic` and `Interpret`.  Instead, you would have to hand-write the `Interpret` instance that loops over the syntax tree, like this:

```haskell
data Node
  = Plain Text
  | Heading { level :: Int, heading :: Text }
  | Listing [Node]

instance Interpret Node where
    autoWith options = Type {..}
      where
        -- Actually, this is not completely correct as you need to handle the more general case where
        -- `"Node"` is α-renamed to another type variable
        expected =
            Pi "Node" Type (Pi _ (Pi _ Text "Node") ...)

        -- Same thing here: you need to handle the more general case where the lamba-bound
        -- variables are renamed, or if any of the names shadow each other
        extract (Lam "Node" _ (Lam "Plain" _ (Lam "Heading" (Lam "Listing" _ e0)))) = buildNode e0
        extract _ = Nothing
          where
            buildNode (App (Var (V "Plain 0)) (TextLit t)) = Just (Plain t)
            ...
            buildNode _ = Nothing
```

The basic idea is that once you skip all the lambdas in the syntax tree it's basically a very verbose encoding of your data type using Dhall's syntax tree and `buildNode` just loops over that and marshals it into the Haskell `Node` type, returning `Nothing` if it runs into something unexpected

**Comment by Profpatsch** [comment]
That’s all very verbose. I definitely wouldn’t have found out how to unmarshal into Haskell.

I suspect the explicit `Recurse` method above would involve a similar manual `Interpret` instance as well?

**Comment by Gabriel439** [comment]
Yes, the explicit `Recurse` method would also require a manual `Interpret` instance, too

**Comment by Profpatsch** [comment]
Hm, do you think it would be possible to write these instances once for any kind of correctly formatted Dhall-expression? e.g. by interpreting into `Fix`?

**Comment by Gabriel439** [comment]
I don't think it's possible mainly due to a limitation in how `GHC.Generics` work.  In fact, you'll notice that the `dhall` library actually already lets you derive an `Interpret` instance for a recursive type without any complaints, but if you try to use the instance the program will crash with a stack overflow

The problem is that `GHC.Generics` has a very simplistic way of dealing with recursion so that if you define a type like:

```haskell
data Stream = Stream { head :: Integer, tail :: Stream }
```

... then the derived `Interpret` implementation will try to decode the following infinite type:

```haskell
{ head : Integer, tail : { head : Integer, tail : { head : Integer, tail : ...
```

As far as I can tell there is no way to do something smarter with the `GHC.Generics` machinery

**Comment by ocharles** [comment]
I just ran into this and would love for this to be at least documented somewhere, perhaps a FAQ?

My use case was wanting to encode what a Nix derivation is, in Dhall. I started with (in `Derivation.dhall`)

```
{
  require : List ./Derivation.dhall
}
```

and got immediately stuck.

**Comment by ocharles** [comment]
Ok, I think it's this:

`Derivation.dhall`:

```
let Derivation : Type =
      forall (Derivation : Type) ->
	  forall (MkDerivation : List Derivation -> Derivation) ->
	  Derivation
in Derivation
```

And `derivation.dhall`:

```
let map = https://ipfs.io/ipfs/QmQ8w5PLcsNz56dMvRtq54vbuPe9cNnCCUXAQp6xLc6Ccx/Prelude/List/map 
in  
let derivation : List ./Derivation.dhall -> ./Derivation.dhall =
      \(inputs : List ./Derivation.dhall) ->
      \(Derivation : Type) ->
      \(MkDerivation : List Derivation -> Derivation) ->
      let f : ./Derivation.dhall -> Derivation = 
            \(x : ./Derivation.dhall) -> x Derivation MkDerivation in
      MkDerivation (map ./Derivation.dhall Derivation f inputs)
in derivation
```

**Comment by Gabriel439** [comment]
@ocharles: Yes, that's correct.  You basically assume that you have the constructor for the recursive type.

I would only suggest a slightly different approach that is a little simpler to encode.  Using a modified version of the example you asked for:

```haskell
data A = MkA { children :: [A], name :: Text }
```

The Dhall encoding of that type would be:

```haskell
∀(A : Type) → ∀(MkA : { children : List A, name : Text } → A) → A
```

And this would be an example value of that type:

```haskell
  λ(A : Type)
→ λ(MkA : { children : List A, name : Text } → A)
→ MkA
  { children =
      [ MkA { children = [] : List A, name = "John" }
      , MkA { children = [] : List A, name = "Mary" }
      ]
  , name     = "Ted"
  }
```

I can add an example like that to the FAQ section

**Comment by ocharles** [comment]
Ah yes, that makes sense. :+1: to a FAQ entry, if we don't get some first class syntax for this.

**Comment by Gabriel439** [comment]
I added a guide on how to do this to the Dhall wiki:

https://github.com/dhall-lang/dhall-lang/wiki/How-to-translate-recursive-code-to-Dhall

... and I also mention this in the Haskell tutorial's FAQ section in this pull request:

https://github.com/dhall-lang/dhall-haskell/pull/223

This is part of an effort to get more documentation migrated to the language-independent Dhall wiki

Once that pull request is merged then I'll mark this closed

**Comment by Gabriel439** [comment]
@mwgkgk: Dhall does not support importing "open" expressions (i.e. expressions with free variables), but there might be another solution to what you have in mind

You can create separate importable files that behave equivalently to the corresponding types and constructors.  The general trick for this is known as Boehm-Berarducci encoding and was created for the purpose of being able to separate each type and term into its own "closed" expression (i.e. an expression with no free variables) while still supporting things like recursion.

Using the example you gave, the way you would split this into multiple files would be:

```haskell
-- ./A

    let A
        : Type
        = ∀(A : Type) → ∀(MkA : { children : List A, name : Text } → A) → A

in  A
```

```haskell
-- ./MkA

    let map =
          https://raw.githubusercontent.com/dhall-lang/Prelude/e44284bc37a5808861dacd4c8bd13d18411cb961/List/map

in  let MkA
        : { children : List ./A, name : Text } → ./A
        =   λ(oldInput : { children : List ./A, name : Text })
          → λ(A : Type)
          → λ(MkA : { children : List A, name : Text } → A)
          →     let adapt : ./A → A = λ(a : ./A) → a A MkA

            in  MkA
                (   { children = map ./A A adapt oldInput.children }
                  ∧ oldInput.{ name }
                )

in  MkA
```

```haskell
-- ./example

    let example : ./A = ./MkA { children = [] : List ./A, name = "John" }

in  example
```

```haskell
-- ./example2

    let example2
        : ./A
        = ./MkA
          { children =
              [ ./MkA { children = [] : List ./A, name = "John" }
              , ./MkA { children = [] : List ./A, name = "Mary" }
              ]
          , name =
              "Ted"
          }

in  example2
```

```bash
$ dhall <<< './example'
∀(A : Type) → ∀(MkA : { children : List A, name : Text } → A) → A

  λ(A : Type)
→ λ(MkA : { children : List A, name : Text } → A)
→ MkA { children = [] : List A, name = "John" }
```

```bash
$ dhall <<< './example2'
∀(A : Type) → ∀(MkA : { children : List A, name : Text } → A) → A

  λ(A : Type)
→ λ(MkA : { children : List A, name : Text } → A)
→ MkA
  { children =
      [ MkA { children = [] : List A, name = "John" }
      , MkA { children = [] : List A, name = "Mary" }
      ]
  , name =
      "Ted"
  }
```

**Comment by Gabriel439** [comment]
@mwgkgk: You might also find these resources helpful, too:

* [Automatic synthesis of typed Λ-programs on term algebras](http://www.sciencedirect.com/science/article/pii/0304397585901355) - The original paper that this is based on
* [Data is Code](http://www.haskellforall.com/2016/04/data-is-code.html) - A post I wrote describing this trick in the context of `morte`/`annah` (the predecessors of `dhall`)

**Comment by jberryman** [comment]
Is there any way (in theory) to hide the complexities of recursive types in some syntactic sugar or language feature? I'm just getting started with dhall and have a really poor intuition for this.

**Comment by Gabriel439** [comment]
@jberryman: There might be some things we could do to make the type-level representation easier to read, but the term-level implementation stuff is going to be difficult to simplify

**Comment by denizdogan** [comment]
Are improvements to this aspect of Dhall on a roadmap somewhere? I don't mean to complain at all, and this is a bit over my head when it comes to type theory, but this is the first time I think "well...that's a bummer."

The smart constructor feature goes a long way for the end-users, but if you're working with complex record structures, the smart constructors themselves can become unreadable to the point where I almost go "you can't do that in Dhall".

So having imagined that this should be possible:

```
let Person = { name : Text, children: List Person }
```

...this doesn't taste quite as good:

```
let Person
    : Type
    = ∀(Person : Type) →
      ∀(MakePerson : { children : List Person, name : Text } → Person) →
        Person

let MakePerson
    : { children : List Person, name : Text } → Person
    = λ(x : { children : List Person, name : Text }) →
      λ(_Person : Type) →
      λ(MakePerson : { children : List _Person, name : Text } → _Person) →
        let adapt
            : Person → _Person
            = λ(y : Person) → y _Person MakePerson

        in  MakePerson
              (x with children = List/map Person _Person adapt x.children)
```

**Comment by kevroletin** [comment]
@Gabriel439
AFAIU in the examples above
* **in dhall** we represented a recursive data structure as code
* **in dhall** we interpreted "data as code" into a flat list
* **in Haskell** we interpreted "data as code" into a recursive tree
 
I am trying to generate a config like the below from dhall. Can I do it without using Haskell just with dhall-to-yaml? The resulting yaml (or json) contains a representation of a tree:
```
- name: John
  children: 
    - name: Mary
      children: []
    - name: Jane
      children: []
```
The imaginary use cases for such configs are:
* build pipelines : one might represent parallel build steps as children of a parent node
* expressing inclusion relationships for events (for example a monitoring tool might use it to produce reports)   

I am very new to dhall but I see that it supports a list (which is a recursive data structure). Maybe there might be a [magic] implementation of a tree data structure?

**Comment by Gabriel439** [comment]
@kevroletin: There is a special tree-like recursive structure which `dhall-to-yaml` supports which is the `Prelude.JSON.Type` type:

https://store.dhall-lang.org/Prelude-v20.2.0/JSON/Type.dhall.html

… so this is one way that you could encode a Dhall data structure that produces that result:

```dhall
-- ./example.dhall

let JSON = https://prelude.dhall-lang.org/JSON/package.dhall

let MakePerson
    : { children : List JSON.Type, name : Text } → JSON.Type
    = λ(x : { children : List JSON.Type, name : Text }) →
        JSON.object
          ( toMap
              { children = JSON.array x.children, name = JSON.string x.name }
          )

let example
    : JSON.Type
    = MakePerson
        { name = "John"
        , children =
          [ MakePerson { name = "Mary", children = [] : List JSON.Type }
          , MakePerson { name = "Jane", children = [] : List JSON.Type }
          ]
        }

in  example
```

```bash
$ dhall-to-yaml --file example.dhall
```
```yaml
children:
  - children: []
    name: Mary
  - children: []
    name: Jane
name: John
```
